### PR TITLE
Skip invalid txs

### PR DESCRIPTION
**Pull request by xiaodino**

### CODE REVIEW

File: bus-mapping/src/circuit_input_builder.rs [hunk 1]
```diff
@@ -550,15 +561,26 @@ impl<P: JsonRpcClient> BuilderClient<P> {
     pub async fn gen_inputs(
```
Comment (xiaodino) [inline bus-mapping/src/circuit_input_builder.rs:561]: gen_inputs will be used by zkevm-chain. Please see the PR: https://github.com/smtmfft/zkevm-chain/pull/2

File: eth-types/src/lib.rs [hunk 2]
```diff
@@ -398,6 +398,9 @@ pub struct ResultGethExecTrace {
 /// before the step is executed.
 #[derive(Deserialize, Serialize, Clone, Debug, Eq, PartialEq)]
 pub struct GethExecTrace {
+    // This one is returned directly by trace_block_by_number in rpc
+    /// True when the tx could not execute
+    pub invalid: Option<bool>,
```
Comment (Brechtpd) [inline eth-types/src/lib.rs:403]: Why is this one an Option? (sorry if this was already the case when I last reviewed this code, I can't remember).

File: bus-mapping/src/circuit_input_builder/transaction.rs [hunk 3]
```diff
@@ -197,6 +197,10 @@ pub struct Transaction {
     pub input: Vec<u8>,
     /// Signature
     pub signature: Signature,
+    /// Invalid tx
+    pub invalid_tx: Option<bool>,
```
Comment (Brechtpd) [inline bus-mapping/src/circuit_input_builder/transaction.rs:201]: Same here, not sure anymore why this is an Option.

**Review by Brechtpd — COMMENTED** [review]
Nice! And not too many changes necessary. :)

File: bus-mapping/src/circuit_input_builder.rs [hunk 4]
```diff
@@ -572,4 +594,26 @@ impl<P: JsonRpcClient> BuilderClient<P> {
...
+        &self,
+        block: &eth_types::Block<eth_types::Transaction>,
+        geth_traces: &mut [eth_types::GethExecTrace],
+        txs_rlp: &eth_types::Bytes,
+    ) {
+        let txs:Vec<eth_types::Transaction> = Rlp::new(txs_rlp).as_list().expect("invalid txs rlp");
+        for (tx, geth_trace) in block.transactions.iter().zip(geth_traces.iter_mut()) {
+            let is_anchor = taiko_anchor::is_anchor(tx);
+            if is_anchor {
+                geth_trace.invalid = Some(false);
```
Comment (smtmfft) [inline bus-mapping/src/circuit_input_builder.rs:609]: I think we can not assume anchor is always valid, although we will have a circuit for anchor later.

File: bus-mapping/src/taiko_anchor.rs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        }],
+        outputs: vec![],
+        #[allow(deprecated)]
+        constant: Some(false),
+        state_mutability: StateMutability::NonPayable,
+    };
+}
+
+/// Check if a tx is a Taiko anchor tx.
+pub fn is_anchor(tx: &eth_types::Transaction) -> bool {
```
Comment (smtmfft) [inline bus-mapping/src/taiko_anchor.rs:27]: you also need to double check this is the 1st tx in the block, maybe you could rename the fn to is_1st_tx_anchor().
Oh, anchor tx has hardcoded from & to, you also need take them into account.

File: bus-mapping/src/circuit_input_builder/transaction.rs (hunk 3 above)
Comment (smtmfft) [inline bus-mapping/src/circuit_input_builder/transaction.rs:201]: Is None type usable somewhere? why is this an option but the other one in another tx type is bool?

File: bus-mapping/src/evm/opcodes.rs [hunk 6]
```diff
@@ -346,12 +346,19 @@ pub fn gen_begin_tx_ops(state: &mut CircuitInputStateRef) -> Result<ExecStep, Er
 
     // Increase caller's nonce
     let caller_address = call.caller_address;
-    let nonce_prev = state.sdb.increase_nonce(&caller_address);
+    let nonce_prev = state.sdb.get_account(&caller_address).1.nonce.as_u64();
+    // Increase caller's nonce when the tx is not invalid
+    let nonce = if !(state.tx.invalid_tx == Some(true)) {
```
Comment (smtmfft) [inline bus-mapping/src/evm/opcodes.rs:351]: != ? 
or maybe if let is better for option check.

File: zkevm-circuits/src/evm_circuit/execution/begin_tx.rs [hunk 7]
```diff
@@ -599,4 +741,132 @@ mod test {
 
         assert_eq!(run_test_circuit_geth_data_default::<Fr>(block), Ok(()));
     }
+
+    #[test]
+    #[should_panic]
+    fn begin_tx_not_enable_skipping_invalid_tx_invalid_nonce() {
```
Comment (smtmfft) [inline zkevm-circuits/src/evm_circuit/execution/begin_tx.rs:747]: The expected panic happens in bus_mapping or in circuit??
For circuit, we need to check all these 4 cases. I think we don't care if a user set the bus_mapping enable flag correctly, as long as the circuit keeps the bottomline safe.

|     tx     | real valid | real invalid | 
| ---------- | -------- | -------- |
| set valid    | positive | negetive|           
| set invalid |       n      |    p         | 

Please make sure your testcases covers all of them in the circuit.

Comment (xiaodino) [reply zkevm-circuits/src/evm_circuit/execution/begin_tx.rs:747]: The expected panic happens in external-tracer. It's neither in bus_mapping nor in circuit.

File: bus-mapping/src/circuit_input_builder.rs [hunk 8]
```diff
@@ -572,4 +594,26 @@ impl<P: JsonRpcClient> BuilderClient<P> {
...
+                continue;
+            }
+            if txs.iter().any(|i| i.hash() == tx.hash) {
+                geth_trace.invalid = Some(false);
+            } else {
+                geth_trace.invalid = Some(true);
+            }
```
[the remark this answers is not in the document]
Comment (xiaodino) [reply bus-mapping/src/circuit_input_builder.rs]: Updated.

File: bus-mapping/src/circuit_input_builder.rs [hunk 9]
```diff
@@ -572,4 +594,26 @@ impl<P: JsonRpcClient> BuilderClient<P> {
...
+        geth_traces: &mut [eth_types::GethExecTrace],
+        txs_rlp: &eth_types::Bytes,
+    ) {
+        let txs:Vec<eth_types::Transaction> = Rlp::new(txs_rlp).as_list().expect("invalid txs rlp");
+        for (tx, geth_trace) in block.transactions.iter().zip(geth_traces.iter_mut()) {
+            let is_anchor = taiko_anchor::is_anchor(tx);
+            if is_anchor {
+                geth_trace.invalid = Some(false);
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (xiaodino) [reply bus-mapping/src/circuit_input_builder.rs]: Removed in_anchor out of the PR.

File: bus-mapping/src/circuit_input_builder/transaction.rs (hunk 3 above)
Comment (xiaodino) [reply bus-mapping/src/circuit_input_builder/transaction.rs]: Removed Option.

File: bus-mapping/src/circuit_input_builder.rs (hunk 4 above)
Comment (xiaodino) [reply bus-mapping/src/circuit_input_builder.rs]: Removed is_anchor.

File: bus-mapping/src/taiko_anchor.rs (hunk 5 above)
Comment (xiaodino) [reply bus-mapping/src/taiko_anchor.rs]: Removed.

File: bus-mapping/src/circuit_input_builder/transaction.rs (hunk 3 above)
Comment (xiaodino) [reply bus-mapping/src/circuit_input_builder/transaction.rs]: Removed Option.

File: bus-mapping/src/evm/opcodes.rs (hunk 6 above)
Comment (xiaodino) [reply bus-mapping/src/evm/opcodes.rs]: Updated.

File: zkevm-circuits/src/evm_circuit/execution/begin_tx.rs (hunk 7 above)
Comment (xiaodino) [reply zkevm-circuits/src/evm_circuit/execution/begin_tx.rs]: Added 4 testcases.

File: bus-mapping/src/circuit_input_builder.rs [hunk 10]
```diff
@@ -550,15 +561,26 @@ impl<P: JsonRpcClient> BuilderClient<P> {
     pub async fn gen_inputs(
         &self,
         block_num: u64,
+        txs_rlp: &eth_types::Bytes,
+        enable_skipping_invalid_tx: bool,
```
[the remark this answers is not in the document]
Comment (xiaodino) [reply bus-mapping/src/circuit_input_builder.rs]: Made it optional.

File: eth-types/src/lib.rs (hunk 2 above)
Comment (xiaodino) [reply eth-types/src/lib.rs]: It should not be an Option. I used Option in Rust in an incorrect way.

File: bus-mapping/src/circuit_input_builder.rs [hunk 11]
```diff
@@ -550,15 +560,28 @@ impl<P: JsonRpcClient> BuilderClient<P> {
     pub async fn gen_inputs(
         &self,
         block_num: u64,
+        txs_rlp: Option<&eth_types::Bytes>,
+        enable_skipping_invalid_tx: bool,
```
Comment (Brechtpd) [inline bus-mapping/src/circuit_input_builder.rs:564]: Do we still need `enable_skipping_invalid_tx`? I assume if you want to pass in the raw transactions using `txs_rlp` it already means some transactions will be missing and so invalid?

File: bus-mapping/src/circuit_input_builder.rs [hunk 12]
```diff
@@ -572,4 +595,21 @@ impl<P: JsonRpcClient> BuilderClient<P> {
...
+        block: &eth_types::Block<eth_types::Transaction>,
+        geth_traces: &mut [eth_types::GethExecTrace],
+        txs_rlp: &eth_types::Bytes,
+    ) {
+        let txs:Vec<eth_types::Transaction> = Rlp::new(txs_rlp).as_list().expect("invalid txs rlp");
+        // Skip the 1st tx which is an anchor tx
+        if block.transactions.len() == 0 && geth_traces.len() == 0 {
+            return
+        }
+        for (tx, geth_trace) in block.transactions[1..].iter().zip(geth_traces[1..].iter_mut()) {
```
Comment (Brechtpd) [inline bus-mapping/src/circuit_input_builder.rs:611]: Is there a reason to skip the anchor transaction here? In theory could of course be marked as invalid but then elsewhere in the anchor specific logic we would just require it to be valid or something like that?

File: eth-types/src/lib.rs [hunk 13]
```diff
@@ -398,6 +398,9 @@ pub struct ResultGethExecTrace {
 /// before the step is executed.
 #[derive(Deserialize, Serialize, Clone, Debug, Eq, PartialEq)]
 pub struct GethExecTrace {
+    /// True when the tx could not execute
+    #[serde(default = "default_bool")]
```
Comment (Brechtpd) [inline eth-types/src/lib.rs:402]: Any specific reason to do `default_bool` for `invalid`? The other values like `failed` don't do it so just wondering.

Comment (xiaodino) [reply eth-types/src/lib.rs:402]: GethExecTrace is being used to deserialize the rpc response of 'debug_traceBlockByNumber'. The response doesn't have invalid. 

If we don't use `#[serde(default = "default_bool")]`, deserialize will fail.

File: bus-mapping/src/circuit_input_builder.rs (hunk 12 above)
Comment (xiaodino) [reply bus-mapping/src/circuit_input_builder.rs:611]: I think we will/should have an anchor circuit just for anchor tx, which will be very different. So here, we can skip anchor tx.

File: bus-mapping/src/circuit_input_builder.rs (hunk 11 above)
Comment (xiaodino) [reply bus-mapping/src/circuit_input_builder.rs]: Good point. Remove `enable_skipping_invalid_tx` here.

**Review by Brechtpd — APPROVED** [review]
I see that there's still some anchor related logic in this PR, which I think is fine for now until we have the anchor implementation.

Otherwise looks good to me!

**Pull request merged**